eth_account/_utils/signing.py

```diff
@@ -44,6 +58,175 @@ def sign_transaction_dict(eth_key, transaction_dict):
+            raise NotImplementedError("TODO: Arrays currently unimplemented in encodeData")
+        else:
+            # First checking to see if the individual values can be encoded
+            try:
+                is_encodable(field["type"], value)
+            except:
+                raise AttributeError(
+                    "Received Invalid type `{0}` of field `{1}`".format(
+                        field["type"],
+                        primaryType,
+                    )
+                )
+
+            # Next see if the data fits the specified encoding type
+            if is_encodable(field["type"], value):
+                # field["type"] is a valid type and this value corresponds to that type.
+                encTypes.append(field["type"])
+                encValues.append(value)
+            else:
+                raise TypeError(
+                    "Value of `{0}` ({2}) of field `{1}` is of the type `{3}`, but expected "
+                    "{4} value".format(
+                        field["name"],
+                        primaryType,
+                        value,
+                        type(value),
+                        field["type"],
+                    )
+                )
+
+    return encode_abi(encTypes, encValues)
+
+
+def validate_structured_data(structured_data):
+    # Check if all the `name` and the `type` attributes in each field of all the
+    # `types` are valid (Regex Check)
+    for field_type in structured_data["types"]:
+        for field in structured_data["types"][field_type]:
+            # Check that field["name"] matches with IDENTIFIER_REGEX
+            if not re.match(IDENTIFIER_REGEX, field["name"]):
+                raise AttributeError(
+                    "Invalid Identifier `{}` in `{}`".format(field["name"], field_type)
+                )
+            # Check that field["type"] matches with TYPE_REGEX
+            if not re.match(TYPE_REGEX, field["type"]):
+                raise AttributeError(
+                    "Invalid Type `{}` in `{}`".format(field["type"], field_type)
+                )
+
+
+def hashStruct(structured_json_string_data, for_domain=False):
+    """
+    The structured_json_string_data is expected to have the ``types`` attribute and
+    the ``primaryType``, ``message``, ``domain`` attribute.
+    The ``for_domain`` variable is used to calculate the ``hashStruct`` as part of the
+    ``domainSeparator`` calculation.
+    """
+    structured_data = json.loads(structured_json_string_data)
+    validate_structured_data(structured_data)
+
+    types = structured_data["types"]
+    if for_domain:
+        primaryType = "EIP712Domain"
+        data = structured_data["domain"]
+    else:
+        primaryType = structured_data["primaryType"]
+        data = structured_data["message"]
+    return keccak(encodeData(primaryType, types, data))
+
+
 # watch here for updates to signature format: https://github.com/ethereum/EIPs/issues/191
```

Comment:
This comment is outdated. I think we can get rid of it. Or change it to EIP712 :)